pkg/trace/stats/span_concentrator.go

```diff
@@ -9,19 +9,106 @@ import (
 	"sync"
 	"time"
 
+	"github.com/DataDog/datadog-agent/pkg/obfuscate"
 	pb "github.com/DataDog/datadog-agent/pkg/proto/pbgo/trace"
 	"github.com/DataDog/datadog-agent/pkg/trace/log"
-	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
 )
 
+// This is a special metric, it's 1 if the span is top-level, 0 if not.
+const topLevelKey = "_top_level"
+
+// measuredKey is a special metric flag that marks a span for trace metrics calculation.
+const measuredKey = "_dd.measured"
+
+// partialVersionKey is a metric carrying the snapshot seq number in the case the span is a partial snapshot
+const partialVersionKey = "_dd.partial_version"
+
 // SpanConcentratorConfig exposes configuration options for a SpanConcentrator
 type SpanConcentratorConfig struct {
 	// ComputeStatsBySpanKind enables/disables the computing of stats based on a span's `span.kind` field
 	ComputeStatsBySpanKind bool
 	// BucketInterval the size of our pre-aggregation per bucket
 	BucketInterval int64
-	// PeerTags additional tags to use for peer entity stats aggregation, nil if disabled
-	PeerTags []string
+}
+
+// StatSpan holds all the required fields from a span needed to calculate stats
+type StatSpan struct {
+	service  string
+	resource string
+	name     string
+	typ      string
+	error    int32
+	parentID uint64
+	start    int64
+	duration int64
+
+	//Fields below this are derived on creation
+
+	spanKind          string
+	statusCode        uint32
+	isTopLevel        bool
+	isMeasured        bool
+	statableSpanKind  bool
```

Comment:
nit: Do we actually need this field on every `StatSpan`? Since the list of _kinds_ for which we can compute stats is static, maybe a separate var or global function can be used instead?

Reply:
I'm wondering if we can pull this logic out of addSpan altogether, and avoid allocating a StatSpan if we're not going to calculate stats on it.

Right now we have (something like)
```
AddSpan(s *StatSpan) {
    if !s.statableSpanKind {
        return
    }
```

Could we invert this and have some function like:
```
func NeedsStats(kind string, meta map[string]string, metrics[string]float64) {
    eligibleSpanKind := sc.computeStatsBySpanKind && computeStatsForSpanKind(kind)
    isTopLevel := metrics[topLevelKey] == 1
    isMeasured := metrics[measuredKey] == 1
    if !(isTopLevel || isMeasured || eligibleSpanKind) {
        return
    }
}
```

And then have clients do:
```
if NeedsStats(s.Kind, s.Meta, s.Metrics) {
    AddSpan(NewStatSpan(...))
}
```

The API isn't as nice, but it would probably save a lot of unnecessary work. WDYT?

Reply:
lol I was actually considering that last night as I was re-thinking the performance implications here (and by avoiding creating the StatSpan on unmeasured spans we could reduce the performance cost here.

One alternative idea I had was if the `NewStatSpan` did "NeedsStats" on its own and just returned `nil` if the provided span didn't need stats? I'm not sure that's better or clearer though

Reply:
I think that's a great idea, but how about we return `nil, false`, i.e.
```
NewStatSpan(...) (s *StatSpan, ok bool) {
...
}
```